src/Core/Types.Sorting/QueryableSortVisitor.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using HotChocolate.Language;
+
+namespace HotChocolate.Types.Sorting
+{
+    public class QueryableSortVisitor
+            : SyntaxNodeVisitor
+    {
+        private readonly ParameterExpression _parameter;
+
+        public QueryableSortVisitor(
+            InputObjectType initialType,
+            Type source)
+        {
+            if (initialType is null)
+            {
+                throw new ArgumentNullException(nameof(initialType));
+            }
+
+            Types.Push(initialType);
+            _parameter = Expression.Parameter(source);
+        }
+
+        protected Queue<SortOperationInvocation> Instance { get; } =
+            new Queue<SortOperationInvocation>();
+
+        protected Stack<IType> Types { get; } =
+            new Stack<IType>();
+
+        public IQueryable<TSource> Sort<TSource>(
+            IQueryable<TSource> source)
+        {
+            if (!Instance.Any())
+            {
+                return source;
+            }
+
+            IOrderedQueryable<TSource> sortedSource
+                = source.AddInitialSortOperation(
+                    Instance.Dequeue(), _parameter);
+
+            while (Instance.Any())
+            {
+                sortedSource
+                    = sortedSource.AddSortOperation(
+                        Instance.Dequeue(), _parameter);
+            }
+
+            return sortedSource;
+        }
+
+        #region Object Value
+
+        public override VisitorAction Enter(
+            ObjectValueNode node,
+            ISyntaxNode parent,
+            IReadOnlyList<object> path,
+            IReadOnlyList<ISyntaxNode> ancestors)
+        {
+            return VisitorAction.Continue;
+        }
+
+        public override VisitorAction Leave(
+            ObjectValueNode node,
+            ISyntaxNode parent,
+            IReadOnlyList<object> path,
+            IReadOnlyList<ISyntaxNode> ancestors)
+        {
+            return VisitorAction.Continue;
+        }
+
+        #endregion
+
+        #region Object Field
+
+        public override VisitorAction Enter(
+            ObjectFieldNode node,
+            ISyntaxNode parent,
+            IReadOnlyList<object> path,
+            IReadOnlyList<ISyntaxNode> ancestors)
+        {
+            if (!(Types.Peek().NamedType() is InputObjectType inputType))
+            {
+                // TODO : resources - invalid type
+                throw new InvalidOperationException();
```

Comment:
NotSupportedException would be better in this case since we will add this later.